Improve balanceTx test coverage for error cases

Anviking opened the pull request:
- [x] Add unit tests for:
    - Unresolved inputs
    - PastHorizonException from validity intervals
    - Bad redeemers
- [x] Tweak comments

### Comments

- See commit history
- Preparation for #3255 

```
Cardano.Wallet.Shelley.Transaction
  balanceTransaction
    [...]
    when passed unresolved inputs
      fails with some error [‐]
        # PENDING: currently silently ignored
      with redeemers
        fails with RedeemerPointsToUnknownScriptHash [✔] (20ms)
    when validity interval is too far in the future
      with some Plutus redeemers
        fails with TimeTranslationPastHorizon [✔] (35ms)
      with no redeemers
        succeeds to balance [✔] (76ms)
    when a redeemer is missing
      balancing succeeds (currently) [✔] (101ms)
    when the input a redeemer is pointing at doesn't exist
      fails with ErrAssignRedeemersTargetNotFound [✔] (35ms)
```

### Issue Number

Preparation for https://input-output.atlassian.net/browse/ADP-1655

Anviking committed:
Improve docs for balanceTx and PartialTx

Anviking committed:
Remove outdated TODO

It was written in October 2021. We have since moved most of the
implementation to Wallet.hs. The remaining logic in Server.hs
is likely -- mostly -- here to stay.

Anviking committed:
Update balanceTx goldens to Babbage txs

Anviking committed:
Move test data for pingPong to top level

Anviking committed:
Improve testing of error cases in balanceTx

- Unresolved inputs
- Validity interval past horizon
- Faulty redeemers

Anviking committed:
Split off distributeSurplusSpec

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Slotting.hs:
```diff
@@ -609,3 +619,32 @@ unsafeExtendSafeZone = f . neverFails r
+        t2 = HF.Bound
+                (RelativeTime 40)
+                (SlotNo 40)
+                (Cardano.EpochNo 2)
```
Note: These values copied from `SlottingSpec.hs`.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Slotting.hs:
```diff
@@ -609,3 +619,32 @@ unsafeExtendSafeZone = f . neverFails r
+-- | A dummy 'TimeInterpreter' for testing. A horizon at @SlotNo 40@ allows
+-- testing 'PastHorizonException's.
+dummyForkInterpreter :: TimeInterpreter (Either PastHorizonException)
+dummyForkInterpreter =
```
Note: the difference between this `dummyForkInterpreter` and the `dummyTimeInterpreter` defined in `TransactionSpec` is that this has a limited horizon, but the `dummyTimeInterpreter` defined in `TransactionSpec` does not.

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2472,8 +2477,113 @@ balanceTransactionSpec = describe "balanceTransaction" $ do
     describe "when passed unresolved inputs" $ do
-        xit "may fail"
-            $ property prop_balanceTransactionUnresolvedInputs
+        it "fails with some error" $ do
```
Note: this applies to a `PartialTx` with one or more inputs that don't appear in the wallet's UTxO set and also don't appear in the explicit list of inputs passed to `balanceTx`.

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2472,8 +2477,113 @@ balanceTransactionSpec = describe "balanceTransaction" $ do
-        xit "may fail"
-            $ property prop_balanceTransactionUnresolvedInputs
+        it "fails with some error" $ do
+            pendingWith "currently silently ignored"
```
Note: this means that `balanceTx` currently silently ignores the condition tested in this test.

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2472,8 +2477,113 @@ balanceTransactionSpec = describe "balanceTransaction" $ do
+                    \got " <> show other
+
+    describe "when the input a redeemer is pointing at doesn't exist" $ do
+        it "fails with TimeTranslationPastHorizon" $ do
```
Should be changed to `ErrAssignRedeemersTargetNotFound`.

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2472,8 +2477,113 @@ balanceTransactionSpec = describe "balanceTransaction" $ do
+            (u, wal, pending)
+            tx
+      where
+        Wallet' u wal pending = mkTestWallet rootK (utxo [Coin 5_000_000])
```
Note: comfortably enough to balance the transaction.

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -4082,6 +4110,61 @@ deserializeBabbageTx :: ByteString -> Cardano.Tx Cardano.BabbageEra
+        { Babbage.txvldt =
+            ValidityInterval
+                (SJust $ SlotNo 1_000_000)
+                (SJust $ SlotNo 2_000_000)
```
Note: these are both far ahead in the future, when compared to the `40` in the `dummyForkInterpreter`.

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1541,8 +1541,17 @@ normalizeDelegationAddress s addr = do
--- The 'inputs' and 'redeemers' must match the binary transaction contained in
--- the 'sealedTx'.
+-- The 'inputs' must match the binary transaction contained in
+-- the 'tx'.
```
Question: should the sets of inputs be identical?

Anviking replied:
Currently: yes - or there'd be undefined behaviour.

With #3255, the `PartialTx.inputs` would be of type `UTxO` (essentially meaning `extraUTxO`), and as long as `(inputs ptx)  <> walletUTxO` is able to resolve all inputs in the tx, we're good. (Meaning: superfluous entries are allowed, and omitting entries which are already in the wallet utxo is ok too)

jonathanknowles replied:
Perhaps the following wording would make the status quo clearer:

The set of 'inputs' must be identical to the set of inputs within the binary transaction contained in the 'tx'.

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1541,8 +1541,17 @@ normalizeDelegationAddress s addr = do
+--
+-- The provided 'redeemers' will overwrite any redeemers inside the 'tx'. This
+-- is done as the internal redeemers in the 'tx' use an index referring to a
+-- 'TxIn', rather than a 'TxIn'. When we are adding extra inputs as part of
```
```suggestion
-- is done as the internal redeemers in the 'tx' use an index referring to a
-- 'TxIn', rather than a 'TxIn' value. When we are adding extra inputs as
-- part of
```

Anviking replied:
Went with `an actual 'TxIn'` instead

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1574,8 +1583,40 @@ balanceTransaction
+    --
+    -- The 'Cardano.EraHistory' and 'SystemStart' is needed to convert
+    -- is needed to convert validity intervals from 'UTCTime' to 'SlotNo' when
+    -- executing Plutus scripts.
```
```suggestion
    -- The 'Cardano.EraHistory' and 'SystemStart' is needed to convert validity
    -- intervals from 'UTCTime' to 'SlotNo' when executing Plutus scripts.
```

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1574,8 +1583,40 @@ balanceTransaction
+    --
+    -- Both values can be hard-coded for a given network configuration. Just be
+    -- cautious that the 'Cardano.EraHistory' will occasionally change as new
+    -- eras are introduced to Cardano. Incorrect 'Cardano.EraHistory' values
```
```suggestion
    -- Both values can be hard-coded for a given network configuration. Just be
    -- cautious of the fact that the 'Cardano.EraHistory' will occasionally
    -- change as new eras are introduced to Cardano.
```

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Slotting.hs:
```diff
@@ -100,7 +103,7 @@ import Control.Tracer
     ( coerce )
 import Data.Functor.Identity
-    ( Identity )
+    ( Identity (runIdentity), runIdentity )
```
```suggestion
    ( Identity (runIdentity) )
```

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Slotting.hs:
```diff
@@ -609,3 +619,32 @@ unsafeExtendSafeZone = f . neverFails r
+            hoistTimeInterpreter (runIdentity . runExceptT) $ mkTimeInterpreter
+                nullTracer
+                (StartTime $ posixSecondsToUTCTime 0)
+                (pure $ HF.mkInterpreter summary)
```
On reflection, I do think that this `dummyForkInterpreter` should be located together with the test code that uses it.

What do you think?

Justification:
- it's never used by the implementation code
- it's only used by test code
- the values used within are essentially arbitrary / hand-picked constants that should never be used in a real implementation.

Anviking replied:
Fair enough! Will give moving a shot.

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2407,73 +2414,178 @@ instance Arbitrary StdGenSeed  where
+                  , TxOut addr (TokenBundle.fromCoin $ Coin 1_000_000)
+                  )
+                | ix <- [0 .. 500]
+                ]
```
Suggestion: I think it could be helpful to write a comment that states the provenance of the hand-picked values in the above code. For example:
- Is it important that `tid = Hash $ B8.replicate 32 '1'`? Could it be `tid = Hash $ B8.replicate 32 '0'` instead?
- Is it important that we have 501 UTxO entries? If so, why is it important? Why not 500, or 100?
- Is it important to choose a 12-byte mnemonic? Could it be a 15-byte mnemonic?

Anviking replied:
(Note that I simply changed the indentation for these tests which are already in master)

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2407,73 +2414,178 @@ instance Arbitrary StdGenSeed  where
+                    [ TxOut addr
+                        (TokenBundle.fromCoin (Coin 100_000_000))
+                    ]
+            totalOutput <$> tx `shouldBe` Right (Coin 102_000_000)
```
Question: where does the extra 2 ada come from?

Anviking replied:
I'm guessing: the change

Anviking replied:
(Note that I simply changed the indentation for these tests which are already in master)

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2407,73 +2414,178 @@ instance Arbitrary StdGenSeed  where
+                    Left (ErrBalanceTxAssignRedeemers
+                            (ErrAssignRedeemersTranslationError
+                                (TimeTranslationPastHorizon
+                                    _pastHoriozon))) -> return ()
```
```suggestion
                    Left
                        (ErrBalanceTxAssignRedeemers
                            (ErrAssignRedeemersTranslationError
                                (TimeTranslationPastHorizon
                                    _pastHoriozon))) -> return ()
```

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2407,73 +2414,178 @@ instance Arbitrary StdGenSeed  where
+                        "Expected pastHorizon failure; got " <> show other
+
+        describe "with no redeemers" $ do
+            it "succeeds to balance" $ do
```
```suggestion
            it "succeeds at balancing" $ do
```

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2407,73 +2414,178 @@ instance Arbitrary StdGenSeed  where
+    describe "when a redeemer is missing" $ do
+        it "balancing succeeds (currently)" $ do
+            -- I would presumably make more sense to fail, if possible, in this
+            -- case, but this test illustrates the current behaviour.
```
```suggestion
            -- It would presumably make more sense to fail, if possible, in this
            -- case, but this test illustrates the current behaviour.
```

Are you able to explain why it would make more sense to fail?

Anviking replied:
```
            -- Plutus script inputs must have a corresponding redeemer. If one
            -- is missing, the resuling tx will not be acceptable by the ledger.
            -- Instead of the current behaviour, it might make more sense to
            -- fail.
            --
            -- The only reason for succeeding might be if a party, say Bob, only
            -- wants to add a redeemer, revealing the redeemer value, after
            -- seeing Alice partially balance the transaction. However it seems
            -- too unclear both whether 1) this is actually is useful, and
            -- 2) whether it would work techincally, aside from lack of
            -- protective guard in balanceTx, so failing might still be saner.

```

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2407,73 +2414,178 @@ instance Arbitrary StdGenSeed  where
+                other -> expectationFailure $
+                    "Expected (Right tx); got " <> show other
+
+    describe "when the input a redeemer is pointing at doesn't exist" $ do
```
```suggestion
    describe "when a redeemer points to an input that doesn't exist" $ do
```

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -4041,33 +4037,133 @@ sealedFee
+        { Babbage.txvldt =
+            ValidityInterval
+                (SJust $ SlotNo 1_000_000)
+                (SJust $ SlotNo 2_000_000)
```
Suggestion: add a comment stating that these are both far ahead in the future, when compared to the slot number of `40` specified in `dummyForkInterpreter`.

Anviking replied:
I collected the magic values more in one place -> 6b5b2ffbb9bbd414fbc5411aa519ea072db99330

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -4041,33 +4037,133 @@ sealedFee
+    body
+        { Babbage.inputs = Babbage.inputs body
+            <> (Set.map (Cardano.toShelleyTxIn . toCardanoTxIn) (Set.fromList ins))
+        }
```
```suggestion
addExtraTxIns ins = modifyBabbageTxBody $ \body -> body
    { Babbage.inputs = Babbage.inputs body
        <> (Set.map (Cardano.toShelleyTxIn . toCardanoTxIn) (Set.fromList ins))
    }
```

Anviking replied:
I do like keeping `body` visually close to the `{`, but can use a `where` helper: 228887eca8b2ed27810d07b8fae098832ccf29bc

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -4041,33 +4037,133 @@ sealedFee
+    :: PartialTx Cardano.BabbageEra
+    -> PartialTx Cardano.BabbageEra
+withValidityBeyondHorizon ptx = ptx
+    { tx = modifyBabbageTxBody setValidityBeyondHorizon (tx ptx)
```
Some strange whitespace character crept in here!

```suggestion
withValidityBeyondHorizon ptx = ptx
    { tx = modifyBabbageTxBody setValidityBeyondHorizon (tx ptx)
```

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -4041,33 +4037,133 @@ sealedFee
+            scriptData
+            auxData
+            scriptValidity)
+        keyWits
```
I tried for a while, but couldn't find a way to make this much shorter!

It would be easy to shorten if this was a record type.

Anviking committed:
Add comment on choice of values in unit tests

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1574,8 +1583,27 @@ balanceTransaction
+    --
+    -- TODO: Remove 'W.ProtocolParameters' argument.
     -> TimeInterpreter (Either PastHorizonException)
+    -- ^ Needed to convert is needed to convert validity intervals from
```
```suggestion
    -- ^ Needed to convert validity intervals from
```

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1574,8 +1583,27 @@ balanceTransaction
     -> ArgGenChange s
     -> (W.ProtocolParameters, Cardano.ProtocolParameters)
+    -- ^ 'Cardano.ProtocolParameters' can be retrived via a Local State Query to
+    -- a local node.
```
```suggestion
    -- ^ 'Cardano.ProtocolParameters' can be retrieved via a Local State Query
    -- to a local node.
```

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -4048,33 +4071,126 @@ sealedFee
+    -- here. Perhaps we could go through cardano-api instead, if we create a
+    -- @TxBodyContent -> Tx@, where - unlike @makeTransactionBody@ - there is
+    -- no validation.
+    alonzoBody = Babbage.TxBody
```
Suggestion: change the identifier to match the era:
```suggestion
    babbageBody = Babbage.TxBody
```

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -4048,33 +4071,126 @@ sealedFee
+        , Babbage.referenceInputs = mempty
+        , Babbage.collateralReturn = SNothing
+        , Babbage.totalCollateral = SNothing
         }
```
Is it possible to re-use [`mkBabbageTxBody`](https://github.com/input-output-hk/cardano-ledger/blob/17649bc09e4c47923f7ad103d337cc1b8e6d3078/eras/babbage/impl/src/Cardano/Ledger/Babbage/TxBody.hs#L940) here?

It seems there's only one field that is non-empty (`outputs`).

Anviking replied:
Ah! Nice, I hadn't seen that.

It seems it's not available in the version of the ledger we're currently using, but I added a note about potentially using it after bumping dependencies.

jonathanknowles reviewed (approved):
Hi @Anviking 

This looks good to me. 👍🏻  Just a few suggestions (see comments).

Anviking committed:
Add note about mkBabbageTxBody

In response to review feedback. It doesn't seem availible with the
ledger version we're currently using.

Anviking committed:
Remove mostly outdated comment about cardano-api
